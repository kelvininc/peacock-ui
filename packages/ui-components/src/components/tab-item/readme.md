# *<kv-tab-item>*



<!-- Auto Generated Below -->


## Usage

### React

```tsx
import React from 'react';
import { KvTabItem } from '@kelvininc/react-ui-components';

export const TabItemExample: React.FC = () => (
	<>
		{/*-- With required props --*/}
		<KvTabItem tabKey="tab1" label="Tab 1" />

		{/*-- Disabled --*/}
		<KvTabItem tabKey="tab1" label="Tab 1" disabled></KvTabItem>

		{/*-- Selected --*/}
		<KvTabItem tabKey="tab1" label="Tab 1" selected></KvTabItem>

		{/*-- With Notification --*/}
		<KvTabItem tabKey="tab1" label="Tab 1" hasNotification></KvTabItem>

		{/*-- With Notification Color --*/}
		<KvTabItem tabKey="tab1" label="Tab 1" hasNotification notificationColor="#202020"></KvTabItem>
	</>
);
```



## Properties

| Property              | Attribute            | Description                                                                                                                                                                             | Type                                           | Default                |
| --------------------- | -------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------- | ---------------------- |
| `customClass`         | `custom-class`       | (optional) Additional classes to apply for custom CSS. If multiple classes are provided they should be separated by spaces. It is also valid to provide CssClassMap with boolean logic. | `CssClassMap \| string \| string[]`            | `''`                   |
| `customStyle`         | --                   | (optional) Additional style to apply for custom CSS.                                                                                                                                    | `{ [key: string]: string; }`                   | `undefined`            |
| `disabled`            | `disabled`           | (optional) To disable this tab                                                                                                                                                          | `boolean`                                      | `false`                |
| `hasNotification`     | `has-notification`   | (optional) To show/hide notification icon or not                                                                                                                                        | `boolean`                                      | `false`                |
| `icon`                | `icon`               | (optional) The tab's icon                                                                                                                                                               | `EIconName \| EOtherIconName`                  | `undefined`            |
| `label` _(required)_  | `label`              | (required) Name to show in UI for this tab                                                                                                                                              | `string`                                       | `undefined`            |
| `notificationColor`   | `notification-color` | (optional) The tab's notification color (hex value, rgb or css var format)                                                                                                              | `string`                                       | `''`                   |
| `selected`            | `selected`           | (optional) To set this tab as the selected one                                                                                                                                          | `boolean`                                      | `false`                |
| `size`                | `size`               | (optional) Sets this tab item to a different styling configuration                                                                                                                      | `EComponentSize.Large \| EComponentSize.Small` | `EComponentSize.Large` |
| `tabKey` _(required)_ | `tab-key`            | (required) A unique identifier for this tab                                                                                                                                             | `number \| string`                             | `undefined`            |


## Events

| Event         | Description                      | Type                            |
| ------------- | -------------------------------- | ------------------------------- |
| `tabSelected` | Emitted when the tab is selected | `CustomEvent<number \| string>` |


## Shadow Parts

| Part     | Description          |
| -------- | -------------------- |
| `"icon"` | The tab's item icon. |


## CSS Custom Properties

| Name                    | Description                       |
| ----------------------- | --------------------------------- |
| `--tab-item-icon-color` | The color of tab navigation icon; |
| `--tab-item-icon-size`  | The size of tab navigation icon;  |


## Dependencies

### Used by

 - [kv-tab-navigation](../tab-navigation)

### Depends on

- [kv-icon](../icon)

### Graph
```mermaid
graph TD;
  kv-tab-item --> kv-icon
  kv-tab-navigation --> kv-tab-item
  style kv-tab-item fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


