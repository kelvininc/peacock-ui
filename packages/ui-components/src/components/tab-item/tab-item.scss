@import '../../assets/styles/globals';

$tab-dot-error-size: 6px;

:host {
	height: inherit;

	.tab-item-container {
		height: 100%;
		display: flex;
		align-items: center;
		margin: 0 $spacing-6x;
		user-select: none;
		white-space: nowrap;
		cursor: pointer;

		&.selected .label {
			color: kv-color('text');
		}

		&.disabled {
			cursor: not-allowed;
			pointer-events: none;

			.label {
				color: kv-color('text', 'step-5');
			}
		}

		.label {
			@include kv-font-h4-semibold();
			color: kv-color('text', 'step-4');
		}

		&.small {
			margin: 0 $spacing-4x;

			.label {
				@include kv-font-label-small-semibold();
			}
		}

		.notification-dot {
			width: $tab-dot-error-size;
			height: $tab-dot-error-size;
			border-radius: 50%;
			margin-left: $spacing-2x;
			background-color: kv-color('error');
			animation: circleGrow 0.25s;
		}
	}

	@keyframes circleGrow {
		from {
			transform: scale(0);
		}

		to {
			transform: scale(1);
		}
	}
}
