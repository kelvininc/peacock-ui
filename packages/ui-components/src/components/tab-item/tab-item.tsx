import { Component, Host, h, Prop, EventEmitter, Event } from '@stencil/core';
import { throttle } from 'lodash-es';
import { EComponentSize } from '../../utils/types';

@Component({
	tag: 'kv-tab-item',
	styleUrl: 'tab-item.scss',
	shadow: true
})
export class KvTabItem {
	/** (required) A unique identifier for this tab */
	@Prop() tabKey!: number | string;
	/** (required) Name to show in UI for this tab */
	@Prop() label!: string;
	/** (optional) To disable this tab */
	@Prop() disabled?: boolean = false;
	/** (optional) To set this tab as the selected one */
	@Prop() selected?: boolean = false;
	/** (optional) To show/hide notification icon or not */
	@Prop() hasNotification: boolean = false;
	/** (optional) The tab's notification color (hex value, rgb or css var format) */
	@Prop() notificationColor?: string = '';
	/** (optional) Sets this tab item to a different styling configuration */
	@Prop() size?: EComponentSize = EComponentSize.Large;
	/** Emitted when the tab is selected */
	@Event() tabSelected: EventEmitter<number | string>;

	private tabClickThrottler: () => void;

	connectedCallback() {
		this.tabClickThrottler = throttle(() => this.tabClick(), 300);
	}

	private tabClick = () => {
		this.tabSelected.emit(this.tabKey);
	};

	render() {
		const customStyles = {
			backgroundColor: this.notificationColor
		};

		return (
			<Host>
				<div
					class={{
						'tab-item-container': true,
						'selected': this.selected,
						'disabled': this.disabled,
						'has-notification': this.hasNotification,
						'small': this.size === EComponentSize.Small
					}}
					onClick={this.tabClickThrottler}
				>
					<div class="label">{this.label}</div>
					{this.hasNotification && <div class="notification-dot" style={customStyles}></div>}
				</div>
			</Host>
		);
	}
}
