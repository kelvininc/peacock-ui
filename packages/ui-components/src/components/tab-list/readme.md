# *<kv-tab-navigation>*



<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- With required props -->
<kv-tab-list selectedTabKey="tab1">
	<kv-tab-item tabKey="tab1" label="Tab 1" />
	<kv-tab-item tabKey="tab2" label="Tab 2" />
	<kv-tab-item tabKey="tab3" label="Tab 3" />
</kv-tab-list>
```


### React

```tsx
import React from 'react';
import { KvTabItem, KvTabList } from '@kelvininc/react-ui-components';

export const TabListExample: React.FC = () => (
	<>
		{/*-- With required props --*/}
		<KvTabList selectedTabKey="tab1">
			<KvTabItem tabKey="tab1" label="Tab 1" />
			<KvTabItem tabKey="tab2" label="Tab 2" />
			<KvTabItem tabKey="tab3" label="Tab 3" />
		</KvTabList>
	</>
);
```



## Properties

| Property                      | Attribute          | Description                                                                                   | Type                                           | Default                |
| ----------------------------- | ------------------ | --------------------------------------------------------------------------------------------- | ---------------------------------------------- | ---------------------- |
| `selectedTabKey` _(required)_ | `selected-tab-key` | (required) The currently selected tab's key (unique identifier)                               | `number \| string`                             | `undefined`            |
| `size`                        | `size`             | (optional) Sets the items on this tab list to use a different styling configuration           | `EComponentSize.Large \| EComponentSize.Small` | `EComponentSize.Large` |
| `tabs` _(required)_           | --                 | (required) The tab items to render in this component to force re-renders when the tabs change | `ITabNavigationItem[]`                         | `undefined`            |


## CSS Custom Properties

| Name                         | Description                  |
| ---------------------------- | ---------------------------- |
| `--tab-list-bg-color`        | Background color of the list |
| `--tab-list-indicator-color` | Tab indicator bar color      |


## Dependencies

### Used by

 - [kv-tab-navigation](../tab-navigation)

### Graph
```mermaid
graph TD;
  kv-tab-navigation --> kv-tab-list
  style kv-tab-list fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


