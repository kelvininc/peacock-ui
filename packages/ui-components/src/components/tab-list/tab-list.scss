@import '../../assets/styles/globals';

$tab-height: 34px;

:host {
	/**
	 * @prop --tab-list-bg-color: Background color of the list
	 * @prop --tab-list-indicator-color: Tab indicator bar color
	 *
	*/
	--tab-list-bg-color: 'transparent';
	--tab-list-indicator-color: #{kv-color('primary')};

	position: relative;
	height: $tab-height;
	width: 100%;
	display: flex;
	background-color: var(--tab-list-bg-color);

	.selected-tab-indicator {
		height: 2px;
		position: absolute;
		bottom: 0;
		transition: all 0.3s ease-out;
		background: var(--tab-list-indicator-color);
	}
}
