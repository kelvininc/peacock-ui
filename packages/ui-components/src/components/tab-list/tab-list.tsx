import { Component, Host, h, State, Prop, Element, Watch } from '@stencil/core';
import { gte, isEmpty } from 'lodash-es';
import { EComponentSize } from '../../utils/types';
import { ITabNavigationItem } from '../tab-navigation/tab-navigation.types';

@Component({
	tag: 'kv-tab-list',
	styleUrl: 'tab-list.scss',
	shadow: true
})
export class KvTabList {
	/** (required) The tab items to render in this component to force re-renders when the tabs change*/
	@Prop() tabs!: ITabNavigationItem[];
	/** (required) The currently selected tab's key (unique identifier) */
	@Prop() selectedTabKey!: number | string;
	/** (optional) Sets the items on this tab list to use a different styling configuration */
	@Prop() size?: EComponentSize = EComponentSize.Large;
	/** Watch for changes on  */
	@Watch('size')
	sizeChangeHandler() {
		this.changeTabsSize();
		this.calculateSelectionAnimationProperties();
	}
	/** Watch for tab selection change and react accordingly by updating the internal states */
	@Watch('selectedTabKey')
	tabSelectionChangeHandler() {
		this.calculateSelectionAnimationProperties();
	}
	/** Watch for tab selection change and react accordingly by updating the internal states */
	@Watch('tabs')
	tabsUpdateChangeHandler() {
		this.tabItems = Array.from(this.el.querySelectorAll('kv-tab-item'));
		this.calculateSelectionAnimationProperties();
	}

	/** The left offset of the tab indicator (in px), updated when the selected tab changes, starts at 24 due to kv-tab-item's margin */
	@State() selectedTabIndicatorOffset: number = 0;
	/** The width of the tab indicator, updated when the selected tab changes */
	@State() selectedTabIndicatorWidth: number = 0;

	// Reference to self
	@Element() el: HTMLKvTabListElement;

	// All the kv-tab-item elements
	private tabItems: HTMLKvTabItemElement[];

	componentDidRender() {
		this.tabItems = Array.from(this.el.querySelectorAll('kv-tab-item'));
		this.changeTabsSize();
		this.calculateSelectionAnimationProperties();
	}

	private changeTabsSize() {
		this.tabItems.forEach(tabItem => (tabItem.size = this.size));
	}

	private calculateSelectionAnimationProperties() {
		const tabIndex = this.tabItems?.findIndex(tab => tab.tabKey === this.selectedTabKey);

		if (gte(tabIndex, 0) && !isEmpty(this.tabItems)) {
			const selectedTabEl = this.tabItems.find((_tabRef, index) => tabIndex === index);
			const labelEl: HTMLElement = selectedTabEl.shadowRoot.querySelector('.label');
			this.selectedTabIndicatorOffset = labelEl.offsetLeft;
			this.selectedTabIndicatorWidth = labelEl.clientWidth / 2;
		}
	}

	render() {
		const tabIndicatorStyle = {
			left: this.selectedTabIndicatorOffset + 'px',
			width: this.selectedTabIndicatorWidth + 'px'
		};

		return (
			<Host>
				<slot></slot>
				<div class="selected-tab-indicator" style={tabIndicatorStyle}></div>
			</Host>
		);
	}
}
