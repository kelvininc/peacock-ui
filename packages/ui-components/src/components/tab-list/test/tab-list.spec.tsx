import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { KvTabItem } from '../../tab-item/tab-item';
import { KvTabList } from '../tab-list';

describe('Tab List (unit tests)', () => {
	let page: SpecPage;
	let component: KvTabList;

	describe('when initialized with required props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvTabList, KvTabItem],
				html: `<kv-tab-list selectedTabKey='dashboard'>
							<kv-tab-item tab-key="dashboard" label="Dashboard"></kv-tab-item>
						</kv-tab-list>
					`
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});
});
