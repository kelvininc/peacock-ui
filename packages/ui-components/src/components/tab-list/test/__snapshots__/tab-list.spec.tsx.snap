// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab List (unit tests) when initialized with required props should match the snapshot 1`] = `
<kv-tab-list selectedtabkey="dashboard">
  <mock:shadow-root>
    <slot></slot>
    <div class="selected-tab-indicator" style="left: 0px; width: 0px;"></div>
  </mock:shadow-root>
  <kv-tab-item label="Dashboard" tab-key="dashboard">
    <mock:shadow-root>
      <div class="tab-item-container">
        <div class="label">
          Dashboard
        </div>
      </div>
    </mock:shadow-root>
  </kv-tab-item>
</kv-tab-list>
`;
